doc/apcaccess.8
.\" manual page [] for apctest version 3.14.6
.Dd March 28, 2014
.Dt APCINFO 8
.Os apcinfo v3.14.6
.Sh NAME
.Nm apcinfo
.Nd retrieve status information from apcupsd(8)
.Sh SYNOPSIS
.Nm 
[-f <config-file>] [-h <host>[:<port>]] [-p <parameter-name>] [-u] [<command>] [<host>[:<port>]]
.Sh DESCRIPTION
.Nm
is a program which prints out the complete status of most American Power 
Conversion Corp. (APC) UPSes provided you have the
.Xr apcupsd 8
daemon installed, properly configured and running. It can access status
information from any APC UPS attached to the localhost or attached to 
any computer on the network which is running apcuspd in NIS mode.
.Pp
The command line options are:
.Bl -hang -width "<command>"
.It -f
configuration file of apcupsd (default is usually /etc/apcupsd/apcupsd.conf)
.It -h 
hostname and port of apcupsd to communicate with.  The default hostname and port are obtained from the configuration file.
.It -p
Report only the value of the named parameter instead of all parameters and values.
.It -u
Remove units field for easier parsing by scripts.
.It <command>
An optional command, unless a hostname is also being specified. The only implemented command is
.Em status\&
Shorthand for
.Em status\& localhost:3551 .
This produces a full printout of the available status information for the 
associated UPS.
.It <host>
An optional hostname which may be a bare machine name, fully qualified 
domain name or IP address.
.It :<port>
An optional port number where a hostname argument has been specified. The 
default is 3551, the official port number assigned by 
.Tn IANA 
for the apcupsd NIS server.  The default hostname and port are obtained from the configuration file.
.El
.Pp
.Nm 
uses apcupsd's inbuilt Network Information Server (NIS) -- no relation 
to Sun's NIS/YP service -- to obtain the current status information from the 
UPS on the local or remote computer. It is therefore
necessary to have the following configuration directives:
.Pp
.Bl -item -compact
.It
        NETSERVER on
.It 
        NISPORT 3551
.El
.Pp
in the
.Pa /etc/apcupsd/apcupsd.conf
configuration file so that the Network Information Server will be accessible.
Firewall rules for appropriate network access may also be needed.
.Sh STATUS REPORT FORMAT
The status report output format is simple ASCII. Generally there
is a single piece of information on each line of output. The
content varies based on the model of UPS being used and, in some cases,
the firmware version. 
.Sh EXAMPLES
Set out below is a sample of the status output which may be
generated for the specified UPS models.
.Pp
.Ss All models
.Bl -hang -width "xxxxxxxxxx" -compact
.It APC
: version, number of records and number of bytes following
.It DATE
: Date and time of last update from UPS
.It HOSTNAME
: hostname of computer running apcupsd
.It VERSION
: apcupsd version number, date and operating system
.It UPSNAME
: UPS name from configuration file (dumb) or EEPROM (smart)
.It CABLE
: Cable type specified in the configuration file
.It MODEL
: UPS model derived from UPS information
.It UPSMODE
: Mode in which UPS is operating
.It STARTTIME
: Date and time apcupsd was started
.It STATUS
: UPS status.  One or more of the following (space-separated):
CAL TRIM BOOST ONLINE ONBATT OVERLOAD LOWBATT REPLACEBATT NOBATT SLAVE SLAVEDOWN
.br
or
.br
COMMLOST
.br
or
.br
SHUTTING DOWN
.It MASTERUPD
: Last time the master sent an update to the slave
.It ENDAPC
: Date and time of status information was written
.El
.Ss Smart-UPS and Matrix-UPS (Smart Signalling)
.Bl -hang -width "xxxxxxxx"-compact
.It LINEV
: Current input line voltage
.It LOADPCT
: Percentage of UPS load capacity used as estimated by UPS
.It BCHARGE
: Current battery capacity charge percentage
.It TIMELEFT
: Remaining runtime left on battery as estimated by the UPS
.It MBATTCHG
: Min battery charge % (BCHARGE) required for system shutdown
.It MINTIMEL
: Min battery runtime (MINUTES) required for system shutdown
.It MAXTIME
: Max battery runtime (TIMEOUT) after which system is shutdown
.It MAXLINEV
: Maximum input line voltage since apcupsd started
.It MINLINEV
: Min (observed) input line voltage since apcupsd started
.It OUTPUTV
: Current UPS output voltage
.It SENSE
: Current UPS sensitivity setting for voltage fluctuations
.It DWAKE
: Time UPS waits after power off when the power is restored
.It DSHUTD
: Delay before UPS powers down after command received
.It DLOWBATT
: Low battery signal sent when this much runtime remains
.It LOTRANS
: Input line voltage below which UPS will switch to battery
.It HITRANS
: Input line voltage above which UPS will switch to battery
.It RETPCT
: Battery charge % required after power off to restore power
.It ITEMP
: UPS internal temperature in degrees Celcius
.It ALARMDEL
: Delay period before UPS starts sounding alarm
.It BATTV
: Current battery voltage
.It LINEFREQ
: Current line frequency in Hertz
.It LASTXFER
: Reason for last transfer to battery since apcupsd startup
.It NUMXFERS
: Number of transfers to battery since apcupsd startup
.It XONBATT
: Date, time of last transfer to battery since apcupsd startup
.It TONBATT
: Seconds currently on battery
.It CUMONBATT
: Cumulative seconds on battery since apcupsd startup
.It XOFFBAT
: Date, time of last transfer off battery since apcupsd startup
.It SELFTEST
: Date and time of last self test since apcupsd startup
.It STESTI
: Self-test interval
.It STATFLAG
: UPS status flag in hex
.It DIPSW
: Current UPS DIP switch settings
.It REG1
: Fault register 1 in hex
.It REG2
: Fault register 2 in hex
.It REG3
: Fault register 3 in hex
.It MANDATE
: UPS date of manufacture
.It SERIALNO
: UPS serial number
.It BATTDATE
: Date battery last replaced (if set)
.It NOMOUTV
: Nominal output voltage to supply when on battery power
.It NOMBATTV
: Nominal battery voltage
.It EXTBATTS
: Number of external batteries (for XL models)
.It BADBATTS
: Number of bad external battery packs (for XL models)
.It FIRMWARE
: UPS firmware version
.It APCMODEL
: APC model information
.El
.Ss Newer Back-UPS Pro (Smart Signalling)
.Bl -hang -width "xxxxxxxx"-compact
.It LINEV
: Current input line voltage
.It LOADPCT
: Percentage of UPS load capacity used
.It MBATTCHG
: Min battery charge % (BCHARGE) required for system shutdown
.It MINTIMEL
: Min battery runtime (MINUTES) required for system shutdown
.It MAXTIME
: Max battery runtime (TIMEOUT) after which system is shutdown
.It MAXLINEV
: Maximum input line voltage since apcupsd startup
.It MINLINEV
: Minimum input line voltage since apcupsd startup
.It OUTPUTV
: Current UPS output voltage
.It BATTV
: Current battery charge voltage
.It STATFLAG
: UPS status flag in hex
.El
.Ss Back-UPS RS 1500 (USB)
.Bl -hang -width "xxxxxxxx"-compact
.It LINEV
: Current input line voltage
.It LOADPCT
: Percentage of UPS load capacity used
.It BCHARGE
: Current battery capacity charge percentage
.It TIMELEFT
: Remaining runtime left on battery as estimated by UPS
.It MBATTCHG
: Min battery charge % (BCHARGE) required for system shutdown
.It MINTIMEL
: Min battery runtime (MINUTES) required for system shutdown
.It MAXTIME
: Max battery runtime (TIMEOUT) after which system is shutdown
.It *OUTPUTV
: Current UPS output voltage
.It *DWAKE
: Time UPS waits after power off when the power is restored
.It *DSHUTD
: Delay before UPS powers down after command received
.It LOTRANS
: Input line voltage below which UPS will switch to battery
.It HITRANS
: Input line voltage above which UPS will switch to battery
.It *RETPCT
: Battery charge % required after power off to restore power
.It *ITEMP
: UPS internal temperature in degrees Celcius
.It ALARMDEL
: Delay period before UPS starts sounding alarm
.It BATTV
: Current battery voltage
.It *LINEFREQ
: Current line frequency in Hertz
.It *LASTXFER
: Reason for last transfer to battery since apcupsd startup
.It NUMXFERS
: Number of transfers to battery since apcupsd startup
.It XONBATT
: Date, time of last transfer to battery since apcupsd startup
.It TONBATT
: Seconds currently on battery
.It CUMONBATT
: Cumulative seconds on battery since apcupsd startup
.It XOFFBAT
: Date, time of last transfer off battery since apcupsd startup
.It SELFTEST
: Date and time of last self test since apcupsd startup
.It STATFLAG
: UPS status flag in hex
.It MANDATE
: UPS date of manufacture
.It SERIALNO
: UPS serial number
.It BATTDATE
: Date battery last replaced (if set)
.It NOMBATTV
: Nominal battery voltage
.It FIRMWARE
: UPS firmware version
.It APCMODEL
: APC model information
.It * presence dependant on USB firmware version.
.El
.Ss Web/SNMP (AP9716) or PowerNet SNMP (AP9605) SmartSlot Card
.Bl -hang -width "xxxxxxxx"-compact
.It LINEV
: Current input line voltage
.It LOADPCT
: Percentage of UPS load capacity used as estimated by UPS
.It BCHARGE
: Current battery capacity charge percentage
.It TIMELEFT
: Remaining runtime left on battery as estimated by UPS
.It MBATTCHG
: Min battery charge % (BCHARGE) required for system shutdown
.It MINTIMEL
: Min battery runtime (MINUTES) required for system shutdown
.It MAXTIME
: Max battery runtime (TIMEOUT) after which system is shutdown
.It MAXLINEV
: Maximum input line voltage since apcupsd startup
.It MINLINEV
: Minimum input line voltage since apcupsd startup
.It OUTPUTV
: UPS output voltage
.It SENSE
: Current UPS sensitivity setting for voltage fluctuations
.It DWAKE
: Time UPS waits after power off when the power is restored
.It DSHUTD
: Delay before UPS powers down after command received
.It DLOWBATT
: Low battery signal sent when this much runtime remains
.It LOTRANS
: Input line voltage below which UPS will switch to battery
.It HITRANS
: Input line voltage above which UPS will switch to battery
.It RETPCT
: Battery charge % required after power off to restore power
.It ITEMP
: UPS internal temperature in degrees Celcius
.It ALARMDEL
: Delay period before UPS starts sounding alarm
.It LINEFREQ
: Current line frequency in Hertz
.It NUMXFERS
: Number of transfers to battery since apcupsd startup
.It TONBATT
: Seconds currently on battery
.It CUMONBATT
: Cumulative seconds on battery since apcupsd startup
.It XOFFBATT
: Date, time of last transfer off battery since apcupsd startup
.It STESTI
: Self-test interval
.It STATFLAG
: UPS status flag in hex
.It DIPSW
: Current UPS DIP switch settings
.It MANDATE
: UPS date of manufacture
.It SERIALNO
: UPS serial number
.It BATTDATE
: Date battery last replaced (if set)
.It NOMOUTV
: Nominal output voltage to supply when on battery power
.It NOMPOWER
: Nominal power output in watts
.It EXTBATTS
: Number of external batteries (for XL models)
.It BADBATTS
: Number of bad external battery packs (for XL models)
.It FIRMWARE
: UPS firmware version
.It APCMODEL
: APC model information
.El
.Ss Share-UPS (AP9270) (Basic port)
.Bl -hang -width "xxxxxxxx"-compact
.It MBATTCHG
: Min battery charge % (BCHARGE) required for system shutdown
.It MINTIMEL
: Min battery runtime (MINUTES) required for system shutdown
.It MAXTIME
: Max battery runtime (TIMEOUT) after which system is shutdown
.It NUMXFERS
: Number of transfers to battery since apcupsd startup
.It TONBATT
: Seconds currently on battery
.It CUMONBATT
: Cumulative seconds on battery since apcupsd startup
.It XOFFBATT
: Date, time of last transfer off battery since apcupsd startup
.It STATFLAG
: UPS status flag in hex
.El
.Ss UPS Environmental Monitoring SmartSlot Card (AP9612TH)
.Bl -hang -width "xxxxxxxx"-compact
.It HUMIDITY
: Ambient humidity
.It AMBTEMP
: Ambient temperature
.El
.Ss Back-UPS Pro and Smart-UPS v/s (Smart Signalling)
.Bl -hang -width "xxxxxxxx"-compact
.It LINEFAIL
: Input line voltage status.
.It BATTSTAT
: Battery status.
.It LASTXFER
: Last (observed) transfer to battery.
.El
.Ss Back-UPS and Net-UPS (Simple Signalling)
.Bl -hang -width "xxxxxxxx"-compact
.It LINEFAIL
: Input line voltage status.
.It BATTSTAT
: Battery status.
.It STATFLAG
: UPS status flag in hex.
.El
.Sh FILES
.Pa /etc/apcupsd/apcupsd.conf
.Sh SEE ALSO
.Xr apcupsd.conf 5 , 
.Xr apcupsd 8 .
.Pp
The HTML apcupsd manual installed on your system or available online at
http://www.apcupsd.org/
.Sh AUTHORS
.Ss This page
.An Trevor Roydhouse
.Ss Software
.An Adam Kropelin (current Project Manager and Code Maintainer)
.An Kern Sibbald (former Project Manager and Code Maintainer)
.An Riccardo Facchetti (former Project Manager and Code Maintainer)
.An Andre M. Hedrick (Project Founder and former Code Maintainer)
.Ss Contributors
An enormous number of users who have devoted their time and energy to
this project -- thanks.


CHANGELOG
# Changelog

## [1.0.3] - 2024 10-08
### Changed
- Renamed apctest to apccmd (apcinfo, apccmd)

### Removed
- Removed events command in daemon (due to useless)

## [1.0.2] - 2024-09-08
### Changed
- Renamed apcaccess to apcinfo (apcinfo, apctest, ...)
- Edited usage help for apcinfo command and removed useless args

## [1.0.1] - 2024-09-08
### Removed
- Examples
- CGI Support
- Gapcmon (GUI libs & implementations)
- SMTP Support
- APC Agent

src/apcaccess.c
/*
 * apcaccess.c
 *
 * Text based IPC management tool for apcupsd package.
 */

/*
 * Copyright (C) 2000-2006 Kern Sibbald
 * Copyright (C) 1996-99 Andre M. Hedrick <andre@suse.com>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of version 2 of the GNU General
 * Public License as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this program; if not, write to the Free
 * Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA 02110-1335, USA.
 */

#include "apc.h"

/* These are all the possible unit labels generated in src/lib/apcstatus.c
 * -u removes them. (To make life easier for scripts.)
 */
const char *const units[] = {
  " Minutes\n",
  " Seconds\n",
  " Percent\n",
  " Volts\n",
  " Watts\n",
  " Hz\n",
  " C\n",
};

/* Behavior modifying flags */
#define NO_UNITS 0x1

/* Get and print status from apcupsd NIS server */
static int do_pthreads_status(const char *host, int port, const char *par, int flags)
{
   sock_t sockfd;
   int n;
   char recvline[MAXSTRING + 1];
   char *line;

   if ((sockfd = net_open(host, NULL, port)) < 0) {
      fprintf(stderr, "Error contacting apcupsd @ %s:%d: %s\n",
         host, port, strerror(-sockfd));
      return 1;
   }

   net_send(sockfd, "status", 6);

   while ((n = net_recv(sockfd, recvline, sizeof(recvline))) > 0) {
      recvline[n] = 0;
      line = recvline;
      if (par) { /* Check for match against parameter name */
         char *r = NULL;
         char *var;

         var = strtok_r(recvline, ":", &r);
         if (!var)   // No : separator? Skip it.
            continue;
         line = recvline + strlen(var) + 1;
         if ((r = strchr(var, ' ')))
            *r = '\0';
         if (strcmp(par, var)) // Doesn't match parameter? Skip it.
            continue;
         while (*line == ' ')
            line++;
      }
      if (flags & NO_UNITS) { /* Remove units labels */
         size_t i;
         for (i = 0; i < sizeof units / sizeof units[0]; i++) {
            const char * const u = units[i];
            size_t ulen = strlen(u);
            size_t llen = strlen(line);

            if (llen >= ulen && !strcmp (line + llen - ulen, u)) {
               line[llen - ulen] = '\n';
               line[llen+1 - ulen] = '\0';
               break;
            }
         }
      }
      fputs(line, stdout);
      // If we had a param to match and we got this far we must have
      // matched it. Clear par to indicate success and bail out.
      if (par) {
         par = NULL;
         break;
      }
   }

   if (n < 0) {
      fprintf(stderr, "Error reading status from apcupsd @ %s:%d: %s\n",
         host, port, strerror(-n));
      net_close(sockfd);
      return 1;
   }

   net_close(sockfd);
   return par ? 2 : 0;
}

/*********************************************************************/

#if defined(HAVE_MINGW)
#undef main
#endif

void usage()
{
   fprintf(stderr, 
      "Usage: apcaccess [-f <config-file>] [-h <host>[:<port>]] "
                       "[-p <parameter-name>] [-u] [<command>] [<host>[:<port>]]\n"
      "\n"
      " -f  Load default host,port from given conf file (default: %s)\n"
      " -h  Connect to host and port (supercedes conf file)\n"
      " -p  Return only the value of the named parameter rather than all parameters and values\n"
      " -u  Strip unit labels\n"
      "\n"
      "Supported commands: 'status' (default)\n"
      "Trailing host/port spec overrides -h and conf file.\n"
      , APCCONF);
}

int main(int argc, char **argv)
{
   const char *par = NULL;
   char *cfgfile = NULL;
   char DEFAULT_HOST[] = "localhost";
   char *host = NULL;
   const char *cmd = "status";
   int port = 3551;
   int flags = 0;
   FILE *cfg;
   UPSINFO ups;

   // Process standard options
   int ch;
   while ((ch = getopt(argc, argv, "f:h:p:u")) != -1)
   {
      switch (ch)
      {
      case 'f':
         cfgfile = optarg;
         break;
      case 'h':
         host = optarg;
         break;
      case 'p':
         par = optarg;
         break;
      case 'u':
         flags |= NO_UNITS;
         break;
      case '?':
      default:
         usage();
         return 1;
      }
   }

   // Default cfgfile if not provided on command line
   // Remember if we defaulted so we know later if conf failure is fatal
   bool fatal = cfgfile != NULL;
   if (!cfgfile)
      cfgfile = APCCONF;

   // Parse conf file for defaults
   if ((cfg = fopen(cfgfile, "r")))
   {
      fclose(cfg);
      memset(&ups, 0, sizeof(UPSINFO));
      init_ups_struct(&ups);
      check_for_config(&ups, cfgfile);
      port = ups.statusport;
      if (!host) // Don't override command line -h
         host = ups.nisip;
   }
   else if (fatal)
   {
      // Failure to find explicitly specified conf file is fatal
      fprintf(stderr, "Unable to open config file '%s'\n", cfgfile);
      return 2;
   }

   // Remaining non-option arguments are optional command and host:port
   // These are from legacy apcaccess syntax.  They take priority over
   // the default from the conf file and/or switch values.
   int optleft = argc - optind;
   if (optleft >= 1)
      cmd = argv[optind];
   if (optleft >= 2)
      host = argv[optind + 1];

   // If still no host, use default
   if (!host || !*host)
      host = DEFAULT_HOST;

   // Separate host and port
   char *p = strchr(host, ':');
   if (p) {
      *p++ = 0;
      port = atoi(p);
   }

   // Translate host of 0.0.0.0 to localhost
   // This is due to NISIP in apcupsd.conf being 0.0.0.0 for listening on all
   // interfaces. In that case just use loopback.
   if (!strcmp(host, "0.0.0.0"))
      host = DEFAULT_HOST;

   if (!strcmp(cmd, "status"))
   {
      return do_pthreads_status(host, port, par, flags);
   }
   else
   {
      fprintf(stderr, "Unknown command %s\n", cmd);
      usage();
      return 1;
   }
}
